HipercowApi/Tools/MetricsUpdateService.cs

```diff
@@ -0,0 +1,181 @@
+        /// maxHoursAgo settings for different states, and the querying can
+        /// be made quicker if it is state specific.
+        /// </summary>
+        /// <param name="cluster">Name of the headnode to query.</param>
+        /// <param name="state">Include only jobs in this state.</param>
+        /// <param name="maxHoursAgo">Include only jobs whose status
+        /// changed within this window.</param>
+        internal void UpdateByState(string cluster, JobState state, int maxHoursAgo)
+        {
+            IScheduler scheduler = this.clusterHandleCache.GetClusterHandle(cluster)!;
+
+            PropertyIdCollection props = [
+                JobPropertyIds.UserName, JobPropertyIds.Owner, JobPropertyIds.ChangeTime,
+                JobPropertyIds.StartTime, JobPropertyIds.EndTime, JobPropertyIds.MinCores,
+                JobPropertyIds.MinNodes, JobPropertyIds.UnitType];
+
+            IFilterCollection jobFilter = scheduler.CreateFilterCollection();
+            jobFilter.Add(FilterOperator.Equal, JobPropertyIds.State, state);
+            ISortCollection sortFilter = new SortCollection
+            {
+                {
+                    SortProperty.SortOrder.Descending,
+                    JobPropertyIds.ChangeTime
+                },
+            };
+
+            ISchedulerRowEnumerator jobs = scheduler.OpenJobEnumerator(
+                props, jobFilter, sortFilter);
+            var now = DateTime.Now;
+            var stateName = (state == JobState.Canceled) ? "Cancelled" : Enum.GetName(state)!;
+            var jobList = jobs.GetRows(int.MaxValue);
+
+            foreach (var job in jobList.Rows)
+            {
+                var changeTime = job[JobPropertyIds.ChangeTime];
+                TimeSpan diff = now - (DateTime)changeTime.Value;
+
+                if (diff.TotalHours >= maxHoursAgo)
+                {
+                    break;
+                }
+
+                string user = Utils.HPCString(job[JobPropertyIds.UserName]);
+                string owner = Utils.HPCString(job[JobPropertyIds.Owner]);
+                user = (user.Trim() == string.Empty) ? owner : user;
+                user = user.Replace("DIDE\\", string.Empty);
+
+                this.userJobs.TryGetValue(user, out Dictionary<string, dynamic>? value);
+                if (value is null)
+                {
+                    value = [];
+                    value["Finished"] = 0;
+                    value["Queued"] = 0;
+                    value["Cancelled"] = 0;
+                    value["Running"] = 0;
+                    value["Failed"] = 0;
+                    value["coreHours"] = 0.0f;
+                    this.userJobs.Add(user, value);
+                }
+
+                value[stateName]++;
+                if ((state == JobState.Finished) || (state == JobState.Running))
+                {
+                    value["coreHours"] += GetCoreHours(job, state);
+                }
+            }
```

Comment:
used multiple times can be made into a private var private static readonly string[] JobStates = ["Finished", "Queued", "Cancelled", "Running", "Failed"];
then can simplify logic to simething like this 
```
        if (!userJobs.TryGetValue(user, out var stats))
            {
                stats = JobStates.ToDictionary(state => state, _ => (dynamic)0);
                stats["coreHours"] = 0.0f;
                userJobs[user] = stats;
            }

            stats[stateName]++;
            if (state is JobState.Finished or JobState.Running)
            {
                stats["coreHours"] += GetCoreHours(job, state);
            }
```

Reply:
OK - will clean that up a bit...